Improved workflow (auto-create directories, quick save to sub directories, move/delete original file)

l3if3rik opened the pull request:
That's basically all the stuff I developed in several branches on my fork. While your program is the best for my purposes, it lacked some features that I really needed.

- Optionally auto-create missing directories
- Optionally show a dialog that allows you to save into one of nine sub directories, just with a button click or key press. I missed that, because only after having selected a cut, I know, if it is some action scene, a dialog, whatever. You can guess from the screenshot that it is especially useful for game recordings.
- Optionally delete or move the original file after the task was processed. If it is moved, the target can be an absolute directory or a sub directory of its original path.

I do not know how to work with translations, so it is all hard-coded English, currently.

To clarify this further, here are some screenshots:

![SimpleVideoCutter_bZj9MTygpA](https://github.com/bartekmotyl/simple-video-cutter/assets/13520695/9062322e-0c90-4985-8863-932dc57ec6c1)

![SimpleVideoCutter_Dy8zstGthE](https://github.com/bartekmotyl/simple-video-cutter/assets/13520695/9aef4fdc-1b20-4f33-aada-1dbbccae9f58)

![SimpleVideoCutter_paemkTt8wX](https://github.com/bartekmotyl/simple-video-cutter/assets/13520695/03a66460-308c-42be-81f7-a7b72172a794)

l3if3rik committed:
Add dialog to save in a specific subdirectory by pressing a digit key

l3if3rik committed:
Autosize quick subdirectory form, disable manual resizing

l3if3rik committed:
Re-add event handlers, rename methods (convention)

l3if3rik committed:
Use KeyPreview, improve Layout

l3if3rik committed:
Change text box background color, change explanation; refactor

l3if3rik committed:
Add functionality to delete or move the original file after the cut was saved

l3if3rik committed:
Assign ActionName to radio buttons tag to reduce redundant string literals

l3if3rik committed:
Change wording

l3if3rik committed:
Add option to create missing directories automatically

l3if3rik committed:
Add buttons to quick sub directory dialog; allow escape key to close

bartekmotyl commented:
Thank you very much for submitting! I am currently a bit busy but will try to review as soon as possible.

bartekmotyl commented on src/SimpleVideoCutter/MainForm.cs:
```diff
@@ -615,21 +596,121 @@ private void StoreSettings()
+                videoViewHover.MediaPlayer.Media?.Dispose();
+                videoViewHover.MediaPlayer.Media = null;
+            }
+        }
```
If such a generic method is added, it should allow to be called in isolation (i.e. without calling `OpenNextFileInDirectory` or `OpenPrevFileInDirectory` which btw. can always fail, as there may be just one file in the directory).  

But it means, that `UnloadVideo` should properly update the whole state (including UI) after unloading the state. 
So, we miss here, for example: 
```
            ClearAllSelections();
            UpdateIndexLabel();
            EnableButtons();
```
as well as clearing the `fileBeingPlayed`

l3if3rik replied:
Probably, UnloadVideo() can be removed completely, as OpenNextFileInDirectory() does all things necessary.  I added the explicit unloading, because I once had a crash, and I noticed that OpenNextFileInDirectory() only implicitly unloaded the video. I didn't like that. Btw, setting fileBeingPlayed to null before calling OpenNextFileInDirectory() would require adjustments to that method. Let's save us the hassle.

bartekmotyl replied:
I would love to keep the `UnloadVideo`, as it would also fix the issue #18. I can imagine that closing a file (to make sure some other program can have exclusive access) may be useful in some workflows. 
I think the changes I suggested would be enough (haven't checked though). 

> btw, setting fileBeingPlayed to null before calling OpenNextFileInDirectory() would require adjustments to that method. Let's save us the hassle.

Yes, currently this would be an issue, but I think adding a `lastFileThatWasPlayed` variable would solve this issue nicely and should be rather straightforward,

l3if3rik replied:
`UnloadVideo()` is ready to be used now.

bartekmotyl commented on src/SimpleVideoCutter/MainForm.cs:
```diff
@@ -30,6 +32,10 @@ public partial class MainForm : Form
         private string? fileToLoadOnStartup = null;
         private Debouncer debouncerHover = new Debouncer();
         private bool playingSelection = false;
+        private bool shouldAskForDeletionConfirmation = true;
```
Why not in `VideoCutterSettings` (as e.g. `KeepSelectionAfterCut` and others) ?

l3if3rik replied:
I like to be reminded once per session. Plus, if that flag was saved in the settings as `false`, we would need an extra checkbox somewhere to set it to `true` again. Then again, why not? Wouldn't be much work.

bartekmotyl replied:
I see your point and of course this is your preference. On the other hand, I would rather like to prefer to keep the consistency and make all the options to work the same way. One could argue ("I like it to be session scope setting") for other variables as well, e.g. `KeepSelectionAfterCut` or `ShowPreview`. If the new settings would behave differently, probably there would be bug reports "Deletion confirmation not saved" etc. 

In general, having possibility to consider all (or some) of the settings as "session based" sound interesting, but I would rather look for a more consistent solution (and I honestly think that in this case global-settings is a better choice). 

> Plus, if that flag was saved in the settings as false, we would need an extra checkbox somewhere to set it to true again. 
Yes, that's true. But I would tend to ignore this. I rather assume that people do not revisit this kind of decisions (have you ever revisited cookie banner setting on a page that you clicked once? 😅  ). So I would assume that possibility to edit the json file (or just delete it and start from scratch) would be a "good enough" workaround.

bartekmotyl commented on src/SimpleVideoCutter/MainForm.cs:
```diff
@@ -30,6 +32,10 @@ public partial class MainForm : Form
         private Debouncer debouncerHover = new Debouncer();
         private bool playingSelection = false;
+        private bool shouldAskForDeletionConfirmation = true;
+        private bool shouldNotifyIfCurrentFileIsBeingDeleted = true;
```
Why not in `VideoCutterSettings` (as e.g. `KeepSelectionAfterCut` and others) ?

l3if3rik replied:
See above.

bartekmotyl replied:
likewise

bartekmotyl commented on src/SimpleVideoCutter/MainForm.cs:
```diff
@@ -726,6 +834,8 @@ private IList<string> GetVideoFilesInDirectory(string currentFilePath)
 
         private string? GetNextPrevFileInDirectory(string currentFilePath, int direction)
         {
+            Debug.WriteLine(currentFilePath);
```
I would rather prefer not to keep `Debug`-s in the code that is in development branch. Feel free to use them in the branch, but remove before opening PR (after testing).

l3if3rik replied:
I think I forgot that line there. Just like a surgeon forgets scissors in the stomach sometimes.

bartekmotyl replied:
Sure, that's not problem. Thanks!

bartekmotyl commented on src/SimpleVideoCutter/MainForm.cs:
```diff
@@ -30,6 +32,10 @@ public partial class MainForm : Form
+        private bool shouldAskForDeletionConfirmation = true;
+        private bool shouldNotifyIfCurrentFileIsBeingDeleted = true;
+
+        public string? FileBeingPlayed { get => fileBeingPlayed; }
```
Tiny nitpick - can be simplified to: 
`public string? FileBeingPlayed  => fileBeingPlayed;`

But in general (see other comments) I think it should not be necessary to expose this information.

bartekmotyl commented on src/SimpleVideoCutter/MainForm.cs:
```diff
@@ -177,12 +183,12 @@ private void MainForm_Load(object? sender, EventArgs e)
                 }
             };
 
+            Globals.PlaceholderFiller = new PlaceholderFiller(videoCutterTimeline1);
```
This feels very wrong to me on many levels. 
First, we  have a static variable elsewhere (in the `Globals`) but we initialize it here.
Second, we push a reference to a _visual_ component to a utility that is responsible for filling placeholders. 
Third: we call it later from `FormSettings` which should be focused only on changing settings - it should not depend on anything else (especially state of the `MainForm` or any other visual components). 

See also the comment in the `FormSettings`.
In general, extracting "replacing placeholders" logic into separate class sounds as very good idea. But still, IMHO the instance of this "replacer" should be owned by the `FormMain` (not a be global), after removing dependency from `FormSettings`, of course.

l3if3rik replied:
> This feels very wrong to me on many levels. 

I wholeheartedly agree. I already removed the functionality of proposing a relative output directory. However, before reading this comment I already refactored PlaceholderFiller, so it does not get videoCutterTimeLine1 in the constructor. My main goal was to remove the message "PlaceholderFiller may be null here", because I know that it cannot be null at the places where it is used.  I know that even the best solution to that has a drawback.
So, since it does not irk you to have FormMain own the instance of PlaceholderFiller (this feels slightly wrong, too), I will adjust my code accordingly.

bartekmotyl commented on src/SimpleVideoCutter/FormSettings.cs:
```diff
@@ -158,12 +176,69 @@ private void buttonOutputDirectory_Click(object sender, EventArgs e)
+                return;
+            }
+
+            string outputDirectory = Globals.PlaceholderFiller.ReplaceStandardDirectoryPatterns(path, mainForm.FileBeingPlayed);
```
I am not sure I understand this approach. As we can see even here, the "replacing placeholders" mechanism depends on the file being played. Why one would like to save anything to setting what is based on (particular) file being played? 
What if my output directory is `{UserVideos}\{FileNameWithoutExtension}` or even `{UserVideos}\{Timestamp}` ?
Does it make sense to show subdirectories of this folder (even if it accidentally exists)?
I would rather consider allowing user to type the subfolder name manually. 
For example, if I usually like to put my cuts into `cuts` subfolder of the folder when main file resides, then I would rather expect user types the `cuts` manually, in the edit box. Alternatively, we can maybe show some standard/suggested names of subfolders (`cuts`, `output`, `svc` etc.) to choose from, but always allow to edit/add a custom one. 

This way we remove dependency on  `PlaceholderFiller` here and the whole logic around `PlaceholderFiller` being global singleton can be avoided (see comment in other `FormMain`)

l3if3rik replied:
See above, I removed that functionality.

bartekmotyl commented on src/SimpleVideoCutter/MainForm.cs:
```diff
@@ -615,21 +596,121 @@ private void StoreSettings()
+
+        private IActionAfterTaskCompletion? GetActionAfterTaskCompletion(FFmpegTask task)
+        {
+            string actionName = VideoCutterSettings.Instance.OriginalFileActionAfterCut ?? "keep";
```
There is a default in `VideoCutterSettings` so I think we do not need a fallback value here.

bartekmotyl commented on src/SimpleVideoCutter/MainForm.cs:
```diff
@@ -615,21 +596,121 @@ private void StoreSettings()
+            string actionName = VideoCutterSettings.Instance.OriginalFileActionAfterCut ?? "keep";
+            IActionAfterTaskCompletion? action = null;
+
+            if (actionName == "keep" || task.InputFilePath == null)
```
Can these actions (implementing `IActionAfterTaskCompletion`) haver a factory (in separate file) to wrap all this code ?
They can even access the `VideoCutterSettings` directly (and maybe even directly modify these settings to store the flag indicating whether to annoy the user with the dialog again). 

Or maybe even better - wouldn't it be enough to assume that these actions have constructors that get `InputFilePath` and need to initialize on their own (access  `VideoCutterSettings` is necessary, or show dialogs/questions to the user). In this case the only thing would be to derive concrete class name from the selected `actionName` (simple convention would do the job here)

l3if3rik replied:
> Or maybe even better - wouldn't it be enough to assume that these actions have constructors that get InputFilePath and need to initialize on their own (access VideoCutterSettings is necessary, or show dialogs/questions to the user). In this case the only thing would be to derive concrete class name from the selected actionName (simple convention would do the job here)

I thought about that, too, and I agree. With the exception that I still would pass the task, not just its `InputFilePath` property, since I want to leave open the possibility of creating diverse actions. My actions just happen to move/delete the input file.

bartekmotyl commented on src/SimpleVideoCutter/VideoCutterSettings.cs:
```diff
@@ -13,6 +14,15 @@ public class VideoCutterSettings
         public string OutputFilePattern { get; set; } = "{FileDate}-{FileNameWithoutExtension}.{Timestamp}{FileExtension}";
+        public string[] QuickSubDirectories { get; set; } = Enumerable.Repeat("", 9).ToArray();
+
+        public bool CreateMissingDirectories { get; set; } = false;
```
I would opt for `true` being the default. This seems to very useful and I do not think it would break any workflows of existing users.

bartekmotyl commented on src/SimpleVideoCutter/MainForm.cs:
```diff
@@ -615,21 +596,121 @@ private void StoreSettings()
+                    confirmDeletionResult = MessageBox.Show(
+                        "As per the settings, the original file will be deleted. Is this OK? If you choose yes, original files will be deleted without another confirmation.",
+                        "Confirm Deletion",
+                        MessageBoxButtons.YesNo
```
I would consider adding also a "cancel" option, that would mean the action is not constructed and the file is not being deleted.

l3if3rik replied:
Good idea!

bartekmotyl commented on src/SimpleVideoCutter/MainForm.cs:
```diff
@@ -658,30 +740,56 @@ private void EnqeueNewTask()
 
-            bool shiftPressed = ModifierKeys == Keys.Shift;
-            if (VideoCutterSettings.Instance.ShowTaskWindow || shiftPressed || !selectionsOnKeyFrames)
+            if (VideoCutterSettings.Instance.ShowTaskWindow || showAddTaskDialog || !selectionsOnKeyFrames)
```
Passing the `showAddTaskDialog` as parameter makes much sense, indeed. 👌

bartekmotyl commented on src/SimpleVideoCutter/MainForm.cs:
```diff
@@ -615,21 +596,121 @@ private void StoreSettings()
+                if (this.shouldNotifyIfCurrentFileIsBeingDeleted)
+                {
+                    DialogResult dialogResult = MessageBox.Show("As per the settings, the original file will be deleted after the cut was saved." +
+                        " Since you are currently playing this file, the program will skip to the next file first." +
```
As a user, I would probably prefer to only unload the current file, without proceeding to the next one automatically (why next? I usually browse file the the reverse date order so I go to previous one). Or at least do not start playing the next file automatically.

l3if3rik replied:
I recently cut about 30 files and enjoyed the jump to the next (I didn't care about the order, could also have been previous) file. Hm, it should be configurable.

bartekmotyl replied:
Yeah, in the end this is really a detail (whether to unload only or load another file immediately, next or prev). If there is a separate option to "close" file then I think it is not that important. 
On the other hand, it can always happen that there is no more files in the folder, thus we end up with "no file loaded" sitaution and this has to be supported.

bartekmotyl commented on src/SimpleVideoCutter/TaskProcessor.cs:
```diff
@@ -189,6 +197,7 @@ public class FFmpegTask
         public FFmpegTaskSelection[]? Selections { get; set; }   
         public long OverallDuration { get; set; }
         public bool Lossless { get; set; }
+        public IActionAfterTaskCompletion? ActionAfterTaskCompletion { get; set; }
```
Have you tested this? 
Currently it fails with exception, because the task is being XML-serialized (cloned) in the `FormAddTask` constructor (I had the `keep` option selected) 

```
System.InvalidOperationException
  HResult=0x80131509
  Message=There was an error reflecting type 'SimpleVideoCutter.FFmpegTask'.
  Source=System.Private.Xml
  StackTrace:
   at System.Xml.Serialization.XmlReflectionImporter.ImportTypeMapping(TypeModel model, String ns, ImportContext context, String dataType, XmlAttributes a, Boolean repeats, Boolean openModel, RecursionLimiter limiter)
   at System.Xml.Serialization.XmlReflectionImporter.ImportElement(TypeModel model, XmlRootAttribute root, String defaultNamespace, RecursionLimiter limiter)
   at System.Xml.Serialization.XmlReflectionImporter.ImportTypeMapping(Type type, XmlRootAttribute root, String defaultNamespace)
   at System.Xml.Serialization.XmlSerializer..ctor(Type type, String defaultNamespace)
   at SimpleVideoCutter.Utils.DeepCloneXML[T](T input) in C:\Users\bartek\Documents\git\simple-video-cutter-mr\src\SimpleVideoCutter\Utils.cs:line 23
   at SimpleVideoCutter.FormAddTask..ctor(FFmpegTask task, Boolean selectionsOnKeyFrames) in C:\Users\bartek\Documents\git\simple-video-cutter-mr\src\SimpleVideoCutter\FormAddTask.cs:line 19
   at SimpleVideoCutter.MainForm.PrepareTask(Boolean showAddTaskDialog) in C:\Users\bartek\Documents\git\simple-video-cutter-mr\src\SimpleVideoCutter\MainForm.cs:line 745
   at SimpleVideoCutter.MainForm.EnqeueNewTask(Boolean showAddTaskDialog) in C:\Users\bartek\Documents\git\simple-video-cutter-mr\src\SimpleVideoCutter\MainForm.cs:line 782
   at SimpleVideoCutter.MainForm.toolStripSelection_ItemClicked(Object sender, ToolStripItemClickedEventArgs e) in C:\Users\bartek\Documents\git\simple-video-cutter-mr\src\SimpleVideoCutter\MainForm.cs:line 1132
   at System.Windows.Forms.ToolStrip.HandleItemClick(ToolStripItem dismissingItem)
   at System.Windows.Forms.ToolStripItem.HandleClick(EventArgs e)
   at System.Windows.Forms.ToolStripItem.HandleMouseUp(MouseEventArgs e)
   at System.Windows.Forms.ToolStrip.OnMouseUp(MouseEventArgs mea)
   at System.Windows.Forms.Control.WmMouseUp(Message& m, MouseButtons button, Int32 clicks)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ScrollableControl.WndProc(Message& m)
   at System.Windows.Forms.ToolStrip.WndProc(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, WM msg, IntPtr wparam, IntPtr lparam)

Inner Exception 1:
InvalidOperationException: Cannot serialize member 'SimpleVideoCutter.FFmpegTask.ActionAfterTaskCompletion' of type 'SimpleVideoCutter.Actions.IActionAfterTaskCompletion', see inner exception for more details.

Inner Exception 2:
NotSupportedException: Cannot serialize member SimpleVideoCutter.FFmpegTask.ActionAfterTaskCompletion of type SimpleVideoCutter.Actions.IActionAfterTaskCompletion because it is an interface.
```

I would consider saving (into the task) only the name of the selected action, and actually instantiate it when it will be about to execute.

l3if3rik replied:
Oh, I never used Shift + E during later development, so I never stumbled upon that. Sorry.

bartekmotyl replied:
This is actually pretty strange, because I think it worked me for the first time I had started the app in your branch (i.e. I was able to cut) but then it failed every time. And btw. I did not use `Shift+E`, I just pressed the `Cut` button.

bartekmotyl commented:
I have take a first look at the changes and posted some comments (I reviewed only a subset of changes so far, but already wanted to give feedback). I like the ideas very much, but I would really prefer to have separate submission for each of the concepts (at least for auto-creating directories, quick-save folders, post-processing actions), that would really simplify the review and allow to integrate some changes whilst still discussing others. 
The first two  (auto-creating directories, quick-save folders) seem very useful and straightforward so I would like to integrate them quickly. 
The third idea (post-processing actions) also looks very good, but there are some conceptual and architectural difficulties (see details in comments I posted). Moreover, currently (at least for me), SVC just crashes when trying to start cutting. 

Anyway, thanks again for taking the effort and posting your changes!

l3if3rik commented:
Thanks for taking the time to review all that! And sorry for the PR containing so much stuff. Initially, I was just tinkering around (it's been at least 10 years since I developed with .NET) and didn't even think of making a PR.

l3if3rik commented:
I will probably address the issues next weekend.

bartekmotyl commented on src/SimpleVideoCutter/ChooseOutputDirectoryDialog.cs:
```diff
@@ -0,0 +1,121 @@
+
+namespace SimpleVideoCutter
+{
+    public partial class ChooseOutputDirectory : Form
```
(I report it here, but this comment applies also to other places) 
I have recently cleaned up development branch from any compiler warnings. In your branch I see 133 warnings. For example here: 

```
Severity	Code	Description	Project	File	Line	Suppression State
Warning (active)	CS8618	Non-nullable field 'originalOutputDir' must contain a non-null value when exiting constructor. Consider declaring the field as nullable.	SimpleVideoCutter	ChooseOutputDirectoryDialog.cs	24	

```

l3if3rik replied:
I cleaned them up in my fork. Though I did not count them, I made much less fixes than 133 (probably between 10 and 20).

The only messages I see now are related to duplicate names in `FormSettings.resx`, and there are 118 of them. Could they have been introduced by my IDE by accident? I did not consciously edit that file.

Excerpt:
```
1>FormSettings.resx : warning MSB3568: Duplicate resource name ">>buttonCancel.Name" is not allowed, ignored.
1>FormSettings.resx : warning MSB3568: Duplicate resource name ">>buttonCancel.Type" is not allowed, ignored.
1>FormSettings.resx : warning MSB3568: Duplicate resource name ">>buttonCancel.Parent" is not allowed, ignored.
1>FormSettings.resx : warning MSB3568: Duplicate resource name ">>buttonCancel.ZOrder" is not allowed, ignored.
...
```

l3if3rik replied:
It turned out that those resource names were there before my changes. Can I safely delete them?

l3if3rik committed:
Use abstract class for actions; allows serialization

l3if3rik committed:
Use "new" keyword

l3if3rik committed:
Follow naming conventions

l3if3rik committed:
Make task readonly

l3if3rik committed:
Handle potential null references

l3if3rik committed:
Prevent PlaceholderFiller being null

l3if3rik committed:
Remove debug logging

l3if3rik committed:
Indicate Windows support

l3if3rik committed:
Declare event handler as nullable

l3if3rik committed:
Adjust wording/variable name

l3if3rik committed:
Use text box instead of combo box

l3if3rik committed:
Remove obsolete property

l3if3rik committed:
Adjust control name

l3if3rik committed:
Move PlaceholderFiller to MainForm

l3if3rik committed:
Save in settings whether there should be a deletion confirmation

l3if3rik committed:
After action executed, do not open next file

l3if3rik committed:
Update GUI after unloading video

l3if3rik committed:
Rename "EnableButtons" to "UpdateButtonStates"

l3if3rik committed:
Fix message

l3if3rik committed:
Remove redundant fallback

l3if3rik committed:
Refactor action system

l3if3rik committed:
Adjust code to conventions

l3if3rik committed:
Store settings when leaving edit mode

l3if3rik committed:
Fix button states after unloading/opening video

l3if3rik commented:
> I would consider adding also a "cancel" option, that would mean the action is not constructed and the file is not being deleted.

Except for this, everything should be resolved now.

`UnloadVideo()` can be used now with an unload/close button that would have to be added.

bartekmotyl commented:
Hi @l3if3rik, just a short update from my side - I have seen you pushed lot of changes, but unfortunately I am very busy recently (plus vacation season) so had no time yet to review, and may still need some time to do it. Sorry for the delay and thanks for you effort!

l3if3rik committed:
Reset file info labels after unloading file

l3if3rik committed:
Fix displayed index of current file

l3if3rik committed:
Determine file index case-insensitively

l3if3rik commented:
https://github.com/bartekmotyl/simple-video-cutter/blob/2618a7cb3fe7da9d6e3b837f976ea615ba34a45a/src/SimpleVideoCutter/MainForm.cs#L775

This bug flew under the radar for quite some time. The files in `videoFileArr` have their native mixed case names. The filename being searched is lowercase. I fixed that by doing a case-insensitive search like in `GetNextPrevFileInDirectory()`.

A method like `GetCurrentFileIndex()` would remove redundancy. It would not have a filename parameter, but use `fileBeingPlayed`. It would fit in, if `GetNextPrevFileInDirectory()` also dropped its argument `currentFilePath` (it's not needed right now).
